parser_tests/regression_test.py

```diff
@@ -0,0 +1,52 @@
+import json
+import os
+from unittest import mock
+
+import pytest
+
+from config import parsers
+
+base_directory = os.path.dirname(os.path.realpath(__file__))
+
+parsers_to_test = [
+    ('aachen', 'academica'),
+]
+
+
+@pytest.mark.parametrize("parser,canteen", parsers_to_test)
+def test_parse_url(parser, canteen):
+    result = parse_mocked(parser, canteen)
+    with open(get_snapshot_result_path(parser, canteen), encoding='utf-8') as result_file:
+        expected_result = result_file.read()
+        assert result == expected_result, ("Actual result:\n"
+                                           + result
+                                           + "\n------ Diff: ------\n")
+
+
+def parse_mocked(parser, canteen):
+    snapshot_website_path = get_snapshot_website_path(parser, canteen)
+    with open(snapshot_website_path) as snapshot_file:
+        website_snapshots = json.load(snapshot_file)
+
+        def mock_response(actual_url):
+            class MockResponse:
+                def read(self):
+                    return website_snapshots[actual_url]
+
+            return MockResponse()
+
+    with mock.patch('urllib.request.urlopen', mock_response):
+        return parsers[parser].parse('', canteen, 'full.xml')
```

Comment:
To be honest, I don't like the whole mocking thing. It just feels very brittle and will also fail in more elaborate cases like if the website requires submitting a form with parameters for displaying the menu.

I think it would be better to just change the parsers to adapt to a simplified test framework. Then every parser could use it however it wants. I haven't come up with a complete design but I have some ideas if you are interested.

Reply:
Yes, definitely! The whole point of this is to just be able to reproduce the answers that a server has given. It is by far not perfect and I'm very intersested in what you have come up with!

Reply:
A simplified version of the leipzig parser looks like this:

```python
def parse_url(url, today):
    canteen = LazyBuilder()
    day = datetime.date.today()
    parse_day(canteen, '{}&date={}'.format(url, day.strftime('%Y-%m-%d')))
    return canteen.toXMLFeed()

def parse_day(canteen, url):
    …
```

This doesn't work well with your mocking approach because `day` depends on the actualy day this code is executed. You would have to mock the complete environment to work around this, which doesn't seem feasable. So I instead suggest that parsers that want to implement regression tests implement two functions (the examples still refers to the leipzig parser).

The first function returns a list of `RegressionState` objects which are serialized and stored by the testing framework. The state is completely custom to the parser and is e.g. stored as a json string:

```python
def generate_state(url):
    day = datetime.date.today().strftime('%Y-%m-%d')
    content = urlopen('{}&date={}'.format(url, day)).read()

    canteen = LazyBuilder()
    feed = parse_day(canteen, content).toXMLFeed()

    return [
        RegressionState(feed, state = content)
    ]
```

The second function is called by the test framework for every stored `RegressionState` and returns the feed which is compared by the test framework.

```python
def test_state(state):
    canteen = LazyBuilder()
    return parse_day(canteen, state).toXMLFeed()
```

Both functions are registered with the `Parser` and called through it:

```python
parser = Parser('leipzig',
    handler = parse_url,
    regression_test = RegressionTest(
        generate = generate_state,
        test = test_state
    ))
```

Reply:
That's great! I also like this way of registering parser to be tested. Will see when I can implement this.

Reply:
I've tried this a bit, but I'm having a lot of trouble with some details. Maybe I can come back to this later, but for now it seems like it's quite a lot of effort.